pkg/rebuild/maven/infer.go

```diff
@@ -57,96 +59,56 @@ func (Rebuilder) CloneRepo(ctx context.Context, t rebuild.Target, repoURI string
 }
 
 func (Rebuilder) InferStrategy(ctx context.Context, t rebuild.Target, mux rebuild.RegistryMux, repoConfig *rebuild.RepoConfig, hint rebuild.Strategy) (rebuild.Strategy, error) {
-	// Do pom.xml search.
 	head, _ := repoConfig.Repository.Head()
 	commitObject, _ := repoConfig.Repository.CommitObject(head.Hash())
-	_, pkgPath, err := findPomXML(commitObject, t.Package)
-	if err != nil {
-		log.Printf("pom.xml path heuristic failed [pkg=%s,repo=%s]: %s\n", t.Package, repoConfig.URI, err.Error())
-	}
 
-	// Do version heuristic search.
-	refMap, err := pomXMLSearch(t.Package, pkgPath, repoConfig.Repository)
+	// TODO: It is possible that the build tool would have changed between the HEAD commit and the commit used for the build.
+	// Although unlikely, we should ideally check out the specific commit used for the build if known.
+	// This would require to do version heuristic first to determine the correct commit/tag.
+	buildTool, err := inferBuildTool(commitObject)
 	if err != nil {
-		log.Printf("pom.xml version heuristic failed [pkg=%s,repo=%s]: %s\n", t.Package, repoConfig.URI, err.Error())
+		return nil, errors.Wrapf(err, "inferring build tool")
 	}
 
-	name, version := t.Package, t.Version
-	cfg := &MavenBuild{}
-	dir := path.Dir(pkgPath)
-	pomXMLGuess := refMap[version]
-	tagGuess, err := rebuild.FindTagMatch(name, version, repoConfig.Repository)
-	if err != nil {
-		return cfg, errors.Wrapf(err, "[INTERNAL] tag heuristic error")
+	switch buildTool {
+	case mavenBuildTool:
+		return MavenInfer(ctx, t, mux, repoConfig)
+	default:
+		return nil, errors.Errorf("unsupported build tool: %s", buildTool)
 	}
+}
 
-	var ref string
-	var commit *object.Commit
-	switch {
-	case tagGuess != "":
-		commit, err = repoConfig.Repository.CommitObject(plumbing.NewHash(tagGuess))
-		if err == nil {
-			if newPath, err := findAndValidatePomXML(commit, name, version, dir); err != nil {
-				log.Printf("registry heuristic tag invalid: %v", err)
-			} else {
-				log.Printf("using tag heuristic ref: %s", tagGuess[:9])
-				ref = tagGuess
-				dir = filepath.Dir(newPath)
-				break
-			}
-		} else if err == plumbing.ErrObjectNotFound {
-			log.Printf("tag heuristic ref not found in repo")
-		} else {
-			return cfg, errors.Wrapf(err, "[INTERNAL] Failed ref resolve from tag [repo=%s,ref=%s]", repoConfig.URI, tagGuess)
+func inferBuildTool(commit *object.Commit) (string, error) {
+	var bestBuildTool string
+	minDepth := math.MaxInt
```

Comment:
can you add a brief explanatory comment saying you prefer prefer a match that's the closest to the repo root?